torch/csrc/utils/future.h

```diff
@@ -47,18 +47,36 @@ class TORCH_API Future final {
       completed_ = true;
       value_ = std::move(value);
 
-      fireCallbacks();
+      std::vector<Callback> cbs;
+      cbs.swap(callbacks_);
+      lock.unlock();
+      // There is no need to protect callbacks_ with the lock.
+      // Once completed_ is set to true, no one can add new callback to the
+      // list. pass value_, error_ for callback to easily check state.
+      for (auto& callback : cbs) {
+        callback(value_, error_.get());
+      }
     }
     finished_cv_.notify_all();
   }
 
   void setError(std::string&& errorMsg) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    AT_ASSERT(!completed());
-    completed_ = true;
-    error_ = c10::guts::make_unique<FutureError>(std::move(errorMsg));
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      TORCH_CHECK(!completed());
+      completed_ = true;
```

Comment:
minor: one other pattern that would be nice to fix here is setting completed_ before the value_ or error_.

I mostly raise this point because we allow completed() to access the atomic<bool> completed_ without holding the lock (which is ok).
This leads to the situation where completed() can momentarily be true, but the value_ is not actually set yet. I think the simple fix is to just set completed_ first
(though I guess this might not unequivocably work with memory reordering, but maybe closer)

Reply:
@jjlilley are you suggesting to set completed_ after value_ and error_? because right now we've already set it first

Reply:
right, that's what I was suggesting.

But it might not matter in practice, the only reason I suggest doing so is because currently, completed() can be accessed without a lock, and can be true, without the variable actually being set.